File: 876.middle-of-the-linked-list/solution2.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun middleNode(head: ListNode?): ListNode? {
        if (head == null) {
            return null
        }
        var faster = head
        var slower = head
        while ((faster != null) && (faster.next != null)) {
            faster = faster.next.next
            slower = slower!!.next
        }
        return slower
    }
}

File: 141.linked-list-cycle/solution2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False
        faster, slower = head, head.next
        if faster is None:
            return False
        faster = faster.next

        while faster:
            faster = faster.next
            if faster:
                faster = faster.next
            if faster is None:
                return False
            
            faster = faster.next
            slower = slower.next
            if faster == slower:
                return True
        return False